geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright 2015 Commonwealth Computer Research, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the License);
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.locationtech.geomesa.kafka
+
+import java.util.Date
+
+import com.typesafe.scalalogging.slf4j.Logging
+import com.vividsolutions.jts.index.quadtree.Quadtree
+import org.geotools.data.store.ContentEntry
+import org.geotools.data.{EmptyFeatureReader, Query}
+import org.geotools.factory.CommonFactoryFinder
+import org.geotools.feature.simple.{SimpleFeatureBuilder, SimpleFeatureTypeBuilder}
+import org.joda.time.{Duration, Instant}
+import org.locationtech.geomesa.filter._
+import org.locationtech.geomesa.kafka.consumer.KafkaConsumerFactory
+import org.locationtech.geomesa.kafka.consumer.KafkaStreamLike.KafkaStreamLikeIterator
+import org.locationtech.geomesa.kafka.consumer.offsets.FindOffset
+import org.locationtech.geomesa.utils.geotools.Conversions._
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes.FR
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter._
+import org.opengis.filter.expression.PropertyName
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+class ReplayKafkaConsumerFeatureSource(entry: ContentEntry,
+                                       replaySFT: SimpleFeatureType,
+                                       liveSFT: SimpleFeatureType,
+                                       topic: String,
+                                       kf: KafkaConsumerFactory,
+                                       replayConfig: ReplayConfig,
+                                       query: Query = null)
+  extends KafkaConsumerFeatureSource(entry, replaySFT, query)
+  with Logging {
+
+  import TimestampFilterSplit.split
+
+  // messages are stored as an array where the most recent is at index 0
+  private[kafka] val messages: Array[GeoMessage] = readMessages()
+
+  override def getReaderForFilter(filter: Filter): FR = {
+
+    val reader: Option[FR] = if (messages.isEmpty) {
+      // no data!
+      None
+    } else {
+      split(filter).flatMap { tfs =>
+        val time = tfs.ts
+        val filter = tfs.filter.getOrElse(Filter.INCLUDE)
+
+        snapshot(time).map(_.getReaderForFilter(filter))
+      }
+    }
+
+    reader.getOrElse(new EmptyFeatureReader[SimpleFeatureType, SimpleFeature](replaySFT))
+  }
+
+  private[kafka] def snapshot(time: Option[Long]): Option[ReplaySnapshotFeatureCache] = {
```

Comment:
I want to be able to unit test the function directly but not make it part of the public api.